### PR TITLE
Revise API for delegating paths

### LINKED ISSUE #567: Revise API for performing delegations
Note: The following issue has been reported by @trishankatdatadog and drafted by @vladimir-v-diaz.

**Description of issue or feature request**:

The top-level Targets role indicates which files can be trusted by clients.  A repository maintainer can delegate full or partial trust of these files to another role.  A delegated role can further delegate these entrusted files to others.  For example, the Targets role can transfer trust of `packages/foo*.tgz` files to role A, and role A can give trust over `packages/foo-1.*.tgz` files to role B.  Note that role A cannot give Role B trust over `packages/foo*.zip` files because the Targets role hasn't granted role A this trust.

The repository tool provides a delegate() method to delegate glob patterns from one role to another.  Unfortunately, the method signature to delegate() can cause users to use it incorrectly.  The API for delegate() is covered [here](https://github.com/theupdateframework/tuf/tree/d4f59e5029b8ef1677150501a944ceb9fb7cd46d/tuf#delegations).  The parameter that indicates the "paths" entrusted to the delegated role is made optional, and a parameter that adds targets to the delegated role is required and listed before the "paths" parameter.

**Current behavior**:

```
repository.targets.delegate(rolename, public_keys, list_of_targets, threshold=1,
    terminating=False, restricted_paths=None, path_hash_prefixes=None)
```

**Expected behavior**:
```
repository.targets.delegate(rolename, public_keys, paths, threshold=1,
    terminating=False, list_of_targets=None, path_hash_prefixes=None)
```

### PR DESCRIPTION
**Pull request by vladimir-v-diaz**
**Fixes issue #**:

Closes #567.

**Description of the changes being introduced by the pull request**:

This pull request revises the API for the delegate() and add_paths() methods of the repository tool.  The documentation on the repo tool is also updated to reflect the changes made to these methods.

The API is revised because the current interface makes the delegated paths of the delegated role an optional argument.  It also uses a name, `restricted_paths`, that is not used in the specification.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Signed-off-by: Vladimir Diaz \<vladimir.v.diaz@gmail.com>

### CODE REVIEW

**Comment by trishankatdatadog** [comment]
Hooray! 🎉

File: tuf/repository_tool.py [hunk 1]
```diff
@@ -2290,8 +2292,8 @@ def delegate(self, rolename, public_keys, list_of_targets, threshold=1,
                 'terminating': terminating,
                 'paths': list(roleinfo['paths'].keys())}
 
-    if restricted_paths is not None:
-      roleinfo['paths'] = relative_restricted_paths
+    if paths is not None:
```
Comment (lukpueh) [inline tuf/repository_tool.py]: `paths` can't be `None`, which means that `roleinfo['paths']` always gets `relative_paths` assigned, which means there is no reason to initialize it with `list(roleinfo['paths'].keys()`.
Also, is it expected that `roleinfo['paths']` can be a dictionary or a list?

File: tuf/repository_tool.py [hunk 2]
```diff
@@ -2122,17 +2123,17 @@ def get_delegated_rolenames(self):
 
 
 
-  def delegate(self, rolename, public_keys, list_of_targets, threshold=1,
-      terminating=False, restricted_paths=None, path_hash_prefixes=None):
+  def delegate(self, rolename, public_keys, paths, threshold=1,
+      terminating=False, list_of_targets=None, path_hash_prefixes=None):
```
Comment (lukpueh) [inline tuf/repository_tool.py]: To me it's not clear why `restricted_paths` has become `paths` and is now mandatory, whereas `list_of_targets` is now optional.

File: tuf/repository_tool.py [hunk 3]
```diff
@@ -1758,32 +1758,33 @@ def target_files(self):
...
+        The child delegation that requires an update to its delegated or
+        trusted paths, as listed in the parent role's delegations (e.g.,
+        'Django' in 'unclaimed').
 
     <Exceptions>
-      securesystemslib.exceptions.Error, if a restricted path in
-      'restricted_paths' is not a string path, or if 'child_rolename' has not
-      been delegated yet.
+      securesystemslib.exceptions.Error, if a delegated path in 'paths' is not
+      a string path, or if 'child_rolename' has not been delegated yet.
```
Comment (lukpueh) [inline tuf/repository_tool.py]: The first condition actually raises a `FormatError`, right? Also, I'd add *"...or not a glob pattern."*. But in reality the `PATHS_SCHEMA` only checks for a list of `AnyStrings` (could be empty too).

File: tuf/repository_tool.py [hunk 4]
```diff
@@ -1809,32 +1810,32 @@ def add_restricted_paths(self, restricted_paths, child_rolename):
...
     for role in roleinfo['delegations']['roles']:
       if role['name'] == child_rolename:
-        restricted_paths = role['paths']
+        delegated_paths = role['paths']
 
     for relative_path in relative_paths:
-      if relative_path not in restricted_paths:
-        restricted_paths.append(relative_path)
+      if relative_path not in delegated_paths:
+        delegated_paths.append(relative_path)
```
Comment (lukpueh) [inline tuf/repository_tool.py]: Could it happen that `delegated_paths` is referenced here before assignment?

Comment (lukpueh) [reply tuf/repository_tool.py]: Also, the only purpose of above two loops seems to be creating the `delegated_paths` list, which isn't used anywhere.

**Pull request merged**